django/db/models/expressions.py

```diff
@@ -925,6 +925,32 @@ def relabeled_clone(self, relabels):
         return self
 
 
+class Excluded(Expression):
```

Comment:
Hi @felixxm, initially, `Excluded` was indeed a subclass of `F`. However, we encountered some limitations, particularly with the SQLite backend. When using operations like `Excluded('rank') + 1`, we found errors, for example, the absence of the `_output_field_or_none` attribute in `Excluded` objects. To solve these problems, `Excluded` was redesigned to inherit directly from `Expression`. This approach enabled us to handle the specifics of excluded values in conflicting update operations, while avoiding `F` inheritance problems."

Reply:
And why you need `output_field`? `Excluded` doesn't depend on the output field.

Reply:
You also documented `Excluded` as an extension of `F` which is currently not True.

Reply:
> And why you need `output_field`? `Excluded` doesn't depend on the output field or uses it.

here's my class as I originally defined it
```
class Excluded(F):
    def __init__(self, field_name):
        super().__init__(field_name)
        if isinstance(field_name, str):
            self.lhs = F(field_name)
        else:
            self.lhs = field_name

    def resolve_expression(
        self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
    ):
        c = self.copy()
        c.lhs = self.lhs.resolve_expression(
            query, allow_joins, reuse, summarize, for_save
        )
        return c

    def as_sql(self, compiler, connection):
        lhs_sql, lhs_params = compiler.compile(self.lhs)
        db_vendor = connection.vendor

        if db_vendor in ["postgresql", "sqlite"]:
            sql = f"EXCLUDED.{lhs_sql.split('.')[-1]}"
        elif db_vendor == "mysql":
            sql = f"VALUES({lhs_sql.split('.')[-1]})"
        else:
            raise NotImplementedError(
                f"Backend '{db_vendor}' is not supported for the Excluded operation."
            )

        return sql, lhs_params

    def copy(self):
        copy = super().copy()
        copy.lhs = self.lhs
        return copy
```
it works for this kind of operation
```
UpsertConflict.objects.bulk_create(
        updated_items,
        update_conflicts=True,
        update_fields=["name", "rank"],
        unique_fields=["number"],
        condition=Q(rank__gt=Excluded("number") ) & ~Q(name__startswith="A") | Q(number__gt=3)
    )
```
but not for
```
UpsertConflict.objects.bulk_create(
        updated_items,
        update_conflicts=True,
        update_fields=["name", "rank"],
        unique_fields=["number"],
        condition=Q(rank__gt=Excluded("number") +1) & ~Q(name__startswith="A") | Q(number__gt=3)
    )
```

Reply:
`F("number") + 1` so `Excluded("number") + 1` should also work when it will be a `F` subclass.

Reply:
> `F("number") + 1` donc `Excluded("number") + 1` devrait également fonctionner lorsqu'il s'agira d'une `F` sous-classe.

yes indeed it should work like that. but it didn't. which pushed me to look for another alternative. i'll try to reproduce the error and provide you with more details on the matter.

Reply:
hello @felixxm , I reproduced the error with Excluded as a subclass of F:
```
class Excluded(F):
    def __init__(self, field_name):
        super().__init__(field_name)
        if isinstance(field_name, str):
            self.lhs = F(field_name)
        else:
            self.lhs = field_name

    def resolve_expression(
        self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
    ):
        c = self.copy()
        c.lhs = self.lhs.resolve_expression(
            query, allow_joins, reuse, summarize, for_save
        )
        return c

    def as_sql(self, compiler, connection):
        lhs_sql, lhs_params = compiler.compile(self.lhs)
        db_vendor = connection.vendor

        if db_vendor in ["postgresql", "sqlite"]:
            sql = f"EXCLUDED.{lhs_sql.split('.')[-1]}"
        elif db_vendor == "mysql":
            sql = f"VALUES({lhs_sql.split('.')[-1]})"
        else:
            raise NotImplementedError(
                f"Backend '{db_vendor}' is not supported for the Excluded operation."
            )

        return sql, lhs_params

    def copy(self):
        copy = super().copy()
        copy.lhs = self.lhs
        return copy
```

```
UpsertConflict.objects.bulk_create(
        updated_items,
        update_conflicts=True,
        update_fields=["name", "rank"],
        unique_fields=["number"],
        condition=Q(rank__gt=Excluded("number") + 1) & ~Q(name__startswith="A") | Q(number__gt=3)
    )
```

```
System check identified no issues (0 silenced).
December 20, 2023 - 10:27:31
Django version 5.1.dev20231218171642, using settings 'test_bulk_create.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.

Internal Server Error: /test/
Traceback (most recent call last):
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/test_bulk_create/myapp/views.py", line 100, in test
    return test_bulk_create_view(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/test_bulk_create/myapp/views.py", line 77, in test_bulk_create_view
    UpsertConflict.objects.bulk_create(
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/query.py", line 847, in bulk_create
    returned_columns = self._batched_insert(
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/query.py", line 1891, in _batched_insert
    self._insert(
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/query.py", line 1862, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 1837, in execute_sql
    for sql, params in self.as_sql():
                       ^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 1780, in as_sql
    ) = self.connection.ops.on_conflict_suffix_sql(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/backends/sqlite3/operations.py", line 440, in on_conflict_suffix_sql
    condition_sql, params = self.compile_bulk_operation_condition_to_sql(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/backends/base/operations.py", line 815, in compile_bulk_operation_condition_to_sql
    sql, params = compiler.compile(node)
                  ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 552, in compile
    sql, params = node.as_sql(self, self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/where.py", line 150, in as_sql
    sql, params = compiler.compile(child)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 552, in compile
    sql, params = node.as_sql(self, self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/where.py", line 150, in as_sql
    sql, params = compiler.compile(child)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 552, in compile
    sql, params = node.as_sql(self, self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/where.py", line 150, in as_sql
    sql, params = compiler.compile(child)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 552, in compile
    sql, params = node.as_sql(self, self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/lookups.py", line 246, in as_sql
    rhs_sql, rhs_params = self.process_rhs(compiler, connection)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/lookups.py", line 428, in process_rhs
    return super().process_rhs(compiler, connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/lookups.py", line 121, in process_rhs
    sql, params = compiler.compile(value)
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/sql/compiler.py", line 550, in compile
    sql, params = vendor_impl(self, self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/expressions.py", line 30, in as_sqlite
    if self.output_field.get_internal_type() == "DecimalField":
       ^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/utils/functional.py", line 47, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/expressions.py", line 307, in output_field
    output_field = self._resolve_output_field()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/barhamou/mes-projets/contribution_a_django/django/django/db/models/expressions.py", line 694, in _resolve_output_field
    type(self.lhs._output_field_or_none),
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Excluded' object has no attribute '_output_field_or_none'
[20/Dec/2023 10:27:33] "GET /test/ HTTP/1.1" 500 194389
```

Reply:
That is complicated :exploding_head: . I tried with really naive implementation (pretty sure it's not bulletproof):
```python
class Excluded(F):
    def as_sql(self, compiler, connection):
        if not connection.features.supports_update_conflicts_with_condition:
            raise NotSupportedError(
                f"{connection.vendor} doesn't support partial updating rows on "
                "conflicts during INSERT"
            )   
        quoted_field_name = compiler.quote_name_unless_alias(self.name)
        sql = f"EXCLUDED.{quoted_field_name}"
        return sql, []
        
    def __repr__(self):
        return f"{self.__class__.__name__}({self.name})"

```
and all test pass :shrug:

Reply:
doesn't pass the test. worse, here's the sql code generated in this debug:thanks for your feedback. with this code the error disappears but:
```
def execute_sql(self, returning_fields=None):
       ...
        with self.connection.cursor() as cursor:
            for sql, params in self.as_sql():
                print('sql=',sql)
                print('params=',params)
                cursor.execute(sql, params)
```
```
sql= INSERT INTO "myapp_upsertconflict" ("number", "rank", "name") VALUES (%s, %s, %s), (%s, %s, %s), (%s, %s, %s), (%s, %s, %s) ON CONFLICT("number") DO UPDATE SET "name" = EXCLUDED."name", "rank" = EXCLUDED."rank" WHERE (("myapp_upsertconflict"."rank" > ("myapp_upsertconflict"."number" + %s) AND NOT ("myapp_upsertconflict"."name" LIKE %s ESCAPE '\')) OR "myapp_upsertconflict"."number" > %s) RETURNING "myapp_upsertconflict"."id"
params= (1, 1, 'Steve', 2, 2, 'Olivia', 3, 3, 'Emma', 4, 4, 'Lucas', 1, 'A%', 3)
[20/Dec/2023 13:38:42] "GET /test/ HTTP/1.1" 200 31
```

Note that EXCLUDE does not appear anywhere in the where clause, but with this class:

```
class Excluded(Expression):
    def __init__(self, field_name):
        super().__init__()
        self.field_name = field_name

    def resolve_expression(self, *args, **kwargs):
        resolved = F(self.field_name).resolve_expression(*args, **kwargs)
        self.output_field = resolved.output_field
        return self

    def as_sql(self, compiler, connection):
        quoted_field_name = compiler.quote_name_unless_alias(self.field_name)
        if connection.vendor in ["postgresql", "sqlite"]:
            sql = f"EXCLUDED.{quoted_field_name}"
        else:
            raise NotSupportedError(
                f"Backend '{connection.vendor}' is not supported for "
                "the Excluded operation."
            )

        return sql, []

    def __repr__(self):
        return f"{self.__class__.__name__}({self.field_name})"
```
we have sql 
```
sql= INSERT INTO "myapp_upsertconflict" ("number", "rank", "name") VALUES (%s, %s, %s), (%s, %s, %s), (%s, %s, %s), (%s, %s, %s) ON CONFLICT("number") DO UPDATE SET "name" = EXCLUDED."name", "rank" = EXCLUDED."rank" WHERE (("myapp_upsertconflict"."rank" > (EXCLUDED."number" + %s) AND NOT ("myapp_upsertconflict"."name" LIKE %s ESCAPE '\')) OR "myapp_upsertconflict"."number" > %s) RETURNING "myapp_upsertconflict"."id"
params= (1, 1, 'Steve', 2, 2, 'Olivia', 3, 3, 'Emma', 4, 4, 'Lucas', 1, 'A%', 3)
[20/Dec/2023 13:51:13] "GET /test/ HTTP/1.1" 200 31
```

Reply:
Then you should start from adding a test coverage for this. A new test should fail with my implementation and pass with your.

It's probably enough to add
```python
    def resolve_expression(self, *args, **kwargs):
        super().resolve_expression(*args, **kwargs)
        return self
```
to my implementation.

Reply:
> Then you should start from adding a test coverage for this. A new test should fail with my implementation and pass with your.
> 
> It's probably enough to add
> 
> ```python
>     def resolve_expression(self, *args, **kwargs):
>         super().resolve_expression(*args, **kwargs)
>         return self
> ```
> 
> to my implementation.

with your code, we're back to square one:
```
class Excluded(F):
    def resolve_expression(self, *args, **kwargs):
        super().resolve_expression(*args, **kwargs)
        return self
    
    def as_sql(self, compiler, connection):
        if not connection.features.supports_update_conflicts_with_condition:
            raise NotSupportedError(
                f"{connection.vendor} doesn't support partial updating rows on "
                "conflicts during INSERT"
            )   
        quoted_field_name = compiler.quote_name_unless_alias(self.name)
        sql = f "EXCLUDED.{quoted_field_name}"
        return sql, []
        
    def __repr__(self):
        return f"{self.__class__.__name__}({self.name})"
```

I remind you that it's on the sqlite backend that the problem lies.
Why do we need Excluded to be a subclass of F? This code solves our problem very well and works perfectly:
```
class Excluded(Expression):
    def __init__(self, field_name):
        super().__init__()
        self.field_name = field_name

    def resolve_expression(self, *args, **kwargs):
        resolved = F(self.field_name).resolve_expression(*args, **kwargs)
        self.output_field = resolved.output_field
        return self

    def as_sql(self, compiler, connection):
        quoted_field_name = compiler.quote_name_unless_alias(self.field_name)
        return connection.ops.get_bulkcreate_condition_sql(quoted_field_name), []

    def __repr__(self):
        return f"{self.__class__.__name__}({self.field_name})"
```

Reply:
> I remind you that it's on the sqlite backend that the problem lies.

Why `Excluded` is not required on PostgreSQL?

> Why do we need Excluded to be a subclass of F?

Because it's natural, the only difference between `F` and `Excluded` is that you want to use `EXCLUDED` instead of a table name.

Reply:
> > I remind you that it's on the sqlite backend that the problem lies.
> 
> Why `Excluded` is not required on PostgreSQL?
> 
> > Why do we need Excluded to be a subclass of F?
> 
> Because it's natural, the only difference between `F` and `Excluded` is that you want to use `EXCLUDED` instead of a table name.

allow me to submit my current work for further review and discussion. i feel there is a lot to be said for the keyword "Excluded".
Besides, I think there must be a serious problem under the hood of F. We'll talk about it later.

Reply:
As a suggestion, you might want to reduce the scope of this PR to getting `condition` into `bulk_create` which supports `F` and `Q` expressions. Then once that's landed, you can create a new PR with adding `Excluded` support.